README.md
[![PyPi](https://pypip.in/v/pvpcbill/badge.svg)](https://pypi.org/project/pvpcbill/)
[![Wheel](https://pypip.in/wheel/pvpcbill/badge.svg)](https://pypi.org/project/pvpcbill/)
[![Travis Status](https://travis-ci.org/azogue/pvpcbill.svg?branch=master)](https://travis-ci.org/azogue/pvpcbill)
[![codecov](https://codecov.io/gh/azogue/pvpcbill/branch/master/graph/badge.svg)](https://codecov.io/gh/azogue/pvpcbill)

# pvpcbill

**Electrical billing simulation for small consumers in Spain using PVPC** (electricity hourly prices).

It uses **[`aiopvpc`](https://github.com/azogue/aiopvpc)** to download PVPC data, and _the usual suspects_ ([`pandas`](https://pandas.pydata.org) & [`matplotlib`](https://matplotlib.org)) to deal with time-series data and plotting.

<span class="badge-buymeacoffee"><a href="https://www.buymeacoffee.com/azogue" title="Donate to this project using Buy Me A Coffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg" alt="Buy Me A Coffee donate button" /></a></span>

## Install

Install from pypi with **`pip install pvpcbill`**, or clone it to run tests or anything else ;-)

## Usage

From a _jupyter notebook_, just call the `create_bill` async helper to instantiate a new 'bill' object:

```python
from pvpcbill import create_bill

# Creación directa de factura
factura = await create_bill(
    path_csv_consumo="/path/to/elec_data/consumo_facturado18_02_2020-18_03_2020-R.csv",
    potencia_contratada=4.6,  # kW
    tipo_peaje="NOC",         # GEN / NOC / VHC
    zona_impuestos="IVA",     # IVA / IGIC / IPSI
)

print(factura)
```

** If using it from a non-async script,
use `asyncio.run(create_bill(**params))` to run the async method.

_Output:_

```text
FACTURA ELÉCTRICA:
--------------------------------------------------------------------------------
* CUPS             	        ES0012345678901234SN
* Fecha inicio             	17/02/2020
* Fecha final              	18/03/2020
* Peaje de acceso          	2.0DHA (Nocturna)
* Potencia contratada      	4.60 kW
* Consumo periodo          	472.93 kWh
* ¿Bono Social?            	No
* Equipo de medida         	0.80 €
* Impuestos                	Península y Baleares (IVA)
* Días facturables         	30
--------------------------------------------------------------------------------

- CÁLCULO DEL TÉRMINO FIJO POR POTENCIA CONTRATADA:
  Peaje acceso potencia:
   4.60 kW x 0.103944 €/kW/día x 30 días (366/2020) = 14.34 €
  Comercialización:
   4.60 kW x 0.008505 €/kW/día x 30 días (366/2020) = 1.17 €
  ==> Término fijo                                                     15.51 €

- CÁLCULO DEL TÉRMINO VARIABLE POR ENERGÍA CONSUMIDA (TARIFA 2.0DHA):
    Periodo 1: 0.111867 €/kWh                             ---> 19.02€(P1)
    - Peaje de acceso:     170 kWh * 0.062012 €/kWh = 10.54€
    - Coste de la energía: 170 kWh * 0.049855 €/kWh = 8.48€
    Periodo 2: 0.045617 €/kWh                             ---> 13.82€(P2)
    - Peaje de acceso:     303 kWh * 0.002215 €/kWh = 0.67€
    - Coste de la energía: 303 kWh * 0.043402 €/kWh = 13.15€
  ==> Término de consumo                                               32.84 €

- IMPUESTO ELÉCTRICO:
    5.11269632% x (15.51€ + 32.84€ = 48.35€)                           2.47 €
  ==> Subtotal                                                         50.82 €

- EQUIPO DE MEDIDA:
    30 días x 0.026667 €/día                                           0.80 €
  ==> Importe total                                                    51.62 €

- IVA O EQUIVALENTE:
    21% de 51.62€                                                      10.84 €

################################################################################
# TOTAL FACTURA                                                        62.46 €
################################################################################
Consumo medio diario en el periodo facturado: 2.08 €/día
```

But there is much more:

```python
# Reparto de costes en la factura
p_imp = (
    + factura.data.termino_impuesto_electrico
    + factura.data.termino_equipo_medida
    + factura.data.termino_iva_total
) / factura.data.total
p_ener = factura.data.termino_variable_total / factura.data.total
p_pot = factura.data.termino_fijo_total / factura.data.total

print(
    f"El coste de la factura se reparte en:\n  "
    f"* un {100*p_ener:.1f} % por energía consumida,\n  "
    f"* un {100*p_pot:.1f} % por potencia contratada,\n  "
    f"* un {100*p_imp:.1f} % por impuestos aplicados\n\n"
)

print(factura.data.to_json())
```

_Output:_

```text
El coste de la factura se reparte en:
  * un 52.6 % por energía consumida,
  * un 24.8 % por potencia contratada,
  * un 22.6 % por impuestos aplicados
```

```json
{
  "config": {
    "tipo_peaje": "NOC",
    "potencia_contratada": 4.6,
    "con_bono_social": false,
    "zona_impuestos": "IVA",
    "alquiler_anual": 9.72,
    "impuesto_electrico": 0.0511269632,
    "cups": "ES0012345678901234SN"
  },
  "num_dias_factura": 30,
  "start": "2020-02-17 00:00:00",
  "end": "2020-03-18 00:00:00",
  "periodos_fact": [
    {
      "billed_days": 30,
      "year": 2020,
      "termino_fijo_peaje_acceso": 14.34,
      "termino_fijo_comercializacion": 1.17,
      "termino_fijo_total": 15.51,
      "energy_periods": [
        {
          "name": "P1",
          "coste_peaje_acceso_tea": 10.544458468,
          "coste_energia_tcu": 8.477372039999999,
          "energia_total": 170.03900000000002
        },
        {
          "name": "P2",
          "coste_peaje_acceso_tea": 0.67090578,
          "coste_energia_tcu": 13.146024950000003,
          "energia_total": 302.892
        }
      ]
    }
  ],
  "descuento_bono_social": 0.0,
  "termino_impuesto_electrico": 2.47,
  "termino_equipo_medida": 0.8,
  "termino_iva_gen": 10.6722,
  "termino_iva_medida": 0.168,
  "termino_iva_total": 10.84,
  "total": 62.46
}
```

### Examples

- [Quick example to simulate a bill (jupyter notebook)](Notebooks/Ejemplo%20rápido.ipynb)
- [Detailed example to simulate a bill (jupyter notebook)](Notebooks/Ejemplo%20simulación%20de%20facturación%20eléctrica%20con%20PVPC.ipynb)


CHANGELOG.md
# Changelog

## [v1.0.0](https://github.com/azogue/pvpcbill/tree/v1.0.0) - Initial (2020-05-08)

**Implemented**

- Helper to parse official energy hourly consumption CSVs
- Helper to download PVPC prices and store them locally
- Helper to simulate an electric bill with consumption + PVPC data
- (de)Serialization of generated dataclasses
- String representation of the electric bill


tests/__init__.py
"""Tests for pvpcbill."""


tests/conftest.py
"""Tests for pvpcbill."""
import json
import pathlib

TEST_EXAMPLES_PATH = pathlib.Path(__file__).parent / "ejemplos_consumo"

TEST_PVPC_STORE = TEST_EXAMPLES_PATH / "pvpc_test_data.csv"
TEST_SAMPLE_1 = TEST_EXAMPLES_PATH / "consumo_facturado18_02_2020-18_03_2020-R.csv"


def load_json_fixture(filename: str):
    """Load stored JSON data."""
    return json.loads((TEST_EXAMPLES_PATH / filename).read_text())


tests/test_billing.py
"""Tests for pvpcbill."""
import pytest

from pvpcbill import create_bill, FacturaData, FacturaElec
from .conftest import load_json_fixture, TEST_PVPC_STORE, TEST_SAMPLE_1


@pytest.mark.parametrize(
    "p_consumption_csv, with_results, power, tariff, tax_zone, discount",
    (
        (TEST_SAMPLE_1, True, 4.6, "NOC", "IVA", False),
        (TEST_SAMPLE_1, True, 4.6, "GEN", "IVA", False),
        # TODO fix VHC tariff, it does not match, should be 62,82 €, not 62.63
        (TEST_SAMPLE_1, False, 4.6, "VHC", "IVA", False),
        (TEST_SAMPLE_1, False, 3.45, "GEN", "IGIC", False),
        (TEST_SAMPLE_1, False, 3.45, "GEN", "IPSI", False),
        (TEST_SAMPLE_1, False, 4.6, "NOC", "IVA", True),
        (TEST_SAMPLE_1, False, 3.45, "NOC", "IGIC", True),
    ),
)
async def test_quick_bill_with_pvpc_store(
    p_consumption_csv, with_results, power, tariff, tax_zone, discount
):
    bill = await create_bill(
        path_csv_consumo=p_consumption_csv,
        path_csv_pvpc_store=TEST_PVPC_STORE,
        potencia_contratada=power,
        tipo_peaje=tariff,
        zona_impuestos=tax_zone,
        con_bono_social=discount,
    )
    assert isinstance(bill, FacturaElec)

    # Check string representation
    str_bill = str(bill)
    assert (
        next(filter(lambda l: l.startswith("# TOTAL FACTURA"), str_bill.splitlines()))
        .rstrip(" €")
        .endswith(str(bill.data.total))
    )
    # print(bill)

    # Check data (de)serialization
    data = bill.to_dict()
    bill_data_2 = FacturaData.from_json(bill.data.to_json())
    assert bill_data_2.to_dict() == data

    # Check results
    res_json_file = f"{bill.data.identifier}.json"
    if with_results:
        ref_results = load_json_fixture(res_json_file)
        assert data == ref_results
    # else:
    #     json_data = bill.data.to_json()
    #     (TEST_EXAMPLES_PATH / res_json_file).write_text(json_data)
    #     print(json_data)


pvpcbill/base.py
# -*- coding: utf-8 -*-
"""Electrical billing for small consumers in Spain using PVPC. Base dataclass."""
import json
from datetime import datetime
from enum import Enum

import attr
import cattr

# Serialization hooks for datetimes and Enums
cattr.register_unstructure_hook(Enum, lambda e: e.value)
cattr.register_structure_hook(Enum, lambda s, enum_cls: enum_cls(s))

time_format = "%Y-%m-%d %H:%M:%S"
cattr.register_unstructure_hook(datetime, lambda dt: dt.strftime(time_format))
cattr.register_structure_hook(datetime, lambda s, _: datetime.strptime(s, time_format))


@attr.s
class Base:
    """
    Base dataclass to store information related to the electric bill.

    * Implements instance methods to serialize the data to dict or JSON
    * And class constructors to
    """

    def to_dict(self):
        """
        Generate dict representation.

        To be stored or shared, so it can be retrieved again with
         `bill = FacturaData.from_dict(data)`
        """
        return cattr.unstructure(self)

    def to_json(self, indent=2, **kwargs) -> str:
        """
        Generate JSON representation.

        To be stored or shared, so it can be retrieved again with
         `bill = FacturaData.from_json(json_data)`
        """
        return json.dumps(self.to_dict(), indent=indent, ensure_ascii=False, **kwargs)

    @classmethod
    def from_dict(cls, raw_data: dict):
        """Constructor from dict representation."""
        return cattr.structure(raw_data, cls)

    @classmethod
    def from_json(cls, raw_data: str):
        """Constructor from JSON representation."""
        return cls.from_dict(json.loads(raw_data))


pvpcbill/models.py
# -*- coding: utf-8 -*-
"""Electrical billing for small consumers in Spain using PVPC. Bill dataclasses."""
from datetime import datetime
from typing import Iterator, List

import attr
import pandas as pd

from pvpcbill.base import Base
from pvpcbill.official import (
    MARGEN_COMERC_EUR_KW_YEAR_MCF,
    round_money,
    round_sum_money,
    split_in_tariff_periods,
    TaxZone,
    TERM_ENER_PEAJE_ACC_EUR_KWH_TEA,
    TERM_POT_PEAJE_ACC_EUR_KW_YEAR_TPA,
    TipoPeaje,
)

DEFAULT_CUPS = "ES00XXXXXXXXXXXXXXDB"
DEFAULT_POTENCIA_CONTRATADA_KW = 3.45
DEFAULT_BONO_SOCIAL = False
DEFAULT_IMPUESTO_ELECTRICO = 0.0511269632  # 4,864% por 1,05113
DEFAULT_ALQUILER_CONT_ANUAL = 0.81 * 12  # € / año para monofásico


@attr.s(auto_attribs=True)
class FacturaConfig(Base):
    """Dataclass to store information related to the electric contract."""

    tipo_peaje: TipoPeaje = attr.ib(default=TipoPeaje.GEN)
    potencia_contratada: float = attr.ib(default=DEFAULT_POTENCIA_CONTRATADA_KW)
    con_bono_social: bool = attr.ib(default=DEFAULT_BONO_SOCIAL)
    zona_impuestos: TaxZone = attr.ib(default=TaxZone.PENINSULA_BALEARES)
    alquiler_anual: float = attr.ib(default=DEFAULT_ALQUILER_CONT_ANUAL)
    impuesto_electrico: float = attr.ib(default=DEFAULT_IMPUESTO_ELECTRICO)
    cups: str = attr.ib(default=DEFAULT_CUPS)


@attr.s(auto_attribs=True)
class EnergykWhTariffPeriod(Base):
    """Dataclass to store info related to the energy in 1 tariff period."""

    name: str = attr.ib()
    coste_peaje_acceso_tea: float = attr.ib(default=0.0)
    coste_energia_tcu: float = attr.ib(default=0.0)
    energia_total: float = attr.ib(default=0.0)


@attr.s(auto_attribs=True)
class FacturaBilledPeriod(Base):
    """Dataclass to store info related to 1 billed period inside a bill."""

    billed_days: int = attr.ib()  # nº de días del periodo facturado
    # TODO cambiar distinción tarifaria de 'año' a periodo reglamentario
    year: int = attr.ib()

    # filled when process hourly consumption
    termino_fijo_peaje_acceso: float = attr.ib(default=0.0)
    termino_fijo_comercializacion: float = attr.ib(default=0.0)
    termino_fijo_total: float = attr.ib(default=0.0)
    energy_periods: List[EnergykWhTariffPeriod] = attr.ib(factory=list)

    @classmethod
    def from_hourly_data(
        cls,
        consumo: pd.Series,
        pvpc_tcu: pd.Series,
        tipo_peaje: TipoPeaje,
        potencia_contratada: float,
    ):
        """
        TODO redo doc
        """
        year = consumo.index[0].year
        billed_days = (consumo.index[-1] - consumo.index[0]).days + 1
        energy_periods = [
            EnergykWhTariffPeriod(
                name=f"P{i+1}",
                coste_peaje_acceso_tea=round_money(cons_period.sum() * coef_tea),
                coste_energia_tcu=round_money(
                    (cons_period * pvpc_tcu.loc[cons_period.index]).sum()
                ),
                energia_total=round_money(cons_period.sum()),
            )
            for i, (coef_tea, cons_period) in enumerate(
                zip(
                    TERM_ENER_PEAJE_ACC_EUR_KWH_TEA[year][tipo_peaje.value],
                    split_in_tariff_periods(consumo, tipo_peaje),
                )
            )
        ]

        billed_period = cls(
            billed_days=billed_days, year=year, energy_periods=energy_periods,
        )

        # Término fijo por peaje de acceso
        billed_period.termino_fijo_peaje_acceso = round_money(
            potencia_contratada * billed_days * billed_period.coef_peaje_acceso_potencia
        )
        # Término fijo por comercialización
        billed_period.termino_fijo_comercializacion = round_money(
            potencia_contratada * billed_days * billed_period.coef_comercializacion
        )
        billed_period.termino_fijo_total = round_money(
            billed_period.termino_fijo_peaje_acceso
            + billed_period.termino_fijo_comercializacion
        )

        return billed_period

    @property
    def total_year_days(self):
        """Total number of days in the billed period's year."""
        return (datetime(self.year + 1, 1, 1) - datetime(self.year, 1, 1)).days

    @property
    def coef_peaje_acceso_potencia(self) -> float:
        coef_y = TERM_POT_PEAJE_ACC_EUR_KW_YEAR_TPA[self.year]
        return round(coef_y / self.total_year_days, 6)

    @property
    def coef_comercializacion(self) -> float:
        coef_y = MARGEN_COMERC_EUR_KW_YEAR_MCF[self.year]
        return round(coef_y / self.total_year_days, 6)


@attr.s(auto_attribs=True)
class FacturaData(Base):
    """Dataclass to store information related to the billed period."""

    config: FacturaConfig = attr.ib()
    num_dias_factura: int = attr.ib()
    start: datetime = attr.ib()
    end: datetime = attr.ib()
    periodos_fact: List[FacturaBilledPeriod] = attr.ib(factory=list)

    descuento_bono_social: float = attr.ib(default=0.0)
    termino_impuesto_electrico: float = attr.ib(default=0.0)
    termino_equipo_medida: float = attr.ib(default=0.0)
    termino_iva_gen: float = attr.ib(default=0.0)
    termino_iva_medida: float = attr.ib(default=0.0)
    termino_iva_total: float = attr.ib(default=0.0)
    total: float = attr.ib(default=0.0)

    def __attrs_post_init__(self):
        """Fill calculated terms of the bill when instantiating."""
        self._calc_taxes_and_total()

    def _calc_taxes_and_total(self):
        """
        Añade los términos finales a la factura eléctrica.

        - Aplica el bono social
        - Calcula el impuesto eléctrico
        - Calcula el coste del alquiler del equipo de medida
        - Añade el IVA y obtiene el total
        """
        subt_fijo_var = self.termino_fijo_total + self.termino_variable_total

        # Cálculo de la bonificación (bono social):
        if self.config.con_bono_social:
            self.descuento_bono_social = round_money(-0.25 * round_money(subt_fijo_var))
            subt_fijo_var += self.descuento_bono_social

        # Cálculo del impuesto eléctrico:
        self.termino_impuesto_electrico = round_money(
            self.config.impuesto_electrico * subt_fijo_var
        )
        subt_fijo_var += self.termino_impuesto_electrico

        # Cálculo del alquiler del equipo de medida:
        frac_year = sum(p.billed_days / p.total_year_days for p in self.periodos_fact)
        self.termino_equipo_medida = round_money(frac_year * self.config.alquiler_anual)

        # Cálculo del IVA y TOTAL:
        self.termino_iva_gen = round_money(
            subt_fijo_var * self.config.zona_impuestos.tax_rate
        )
        self.termino_iva_medida = round_money(
            self.termino_equipo_medida * self.config.zona_impuestos.measurement_tax_rate
        )
        self.termino_iva_total = round_money(
            self.termino_iva_gen + self.termino_iva_medida
        )

        # TOTAL FACTURA:
        subt_fijo_var += self.termino_equipo_medida + self.termino_iva_total
        self.total = round_money(subt_fijo_var)

    def iter_energy_periods(self) -> Iterator[EnergykWhTariffPeriod]:
        """Itera sobre cada periodo tarifario de cada periodo de facturación."""
        for billed_period in self.periodos_fact:
            for ener_period in billed_period.energy_periods:
                yield ener_period

    @property
    def coste_total_peaje_acceso_tea(self) -> float:
        return round_sum_money(
            ener_p.coste_peaje_acceso_tea for ener_p in self.iter_energy_periods()
        )

    @property
    def coste_total_energia_tcu(self) -> float:
        return round_sum_money(
            ener_p.coste_energia_tcu for ener_p in self.iter_energy_periods()
        )

    @property
    def consumo_total(self) -> float:
        """Calcula la energía total facturada, sumando cada periodo de facturación."""
        return sum(ener_p.energia_total for ener_p in self.iter_energy_periods())

    @property
    def termino_variable_total(self) -> float:
        """Calcula el coste por energía total, sumando cada periodo de facturación."""
        return round_money(
            self.coste_total_peaje_acceso_tea + self.coste_total_energia_tcu
        )

    @property
    def termino_fijo_total(self) -> float:
        """Calcula el coste por potencia total, sumando cada periodo de facturación."""
        return round_sum_money(
            billed_period.termino_fijo_total for billed_period in self.periodos_fact
        )

    @property
    def identifier(self) -> str:
        """Text identifier to be used as filename for exported data."""
        str_ident = f"elecbill_data_{self.start:%Y_%m_%d}_to_{self.end:%Y_%m_%d}_"
        str_ident += f"{self.config.tipo_peaje.value}_"
        str_ident += f"{self.config.potencia_contratada:g}_".replace(".", "_")
        str_ident += f"{self.config.zona_impuestos.value}"
        if self.config.con_bono_social:
            str_ident += f"_discount"
        return str_ident


pvpcbill/handler.py
# -*- coding: utf-8 -*-
"""Electrical billing for small consumers in Spain using PVPC. Bill handler."""
import pandas as pd

from pvpcbill.models import FacturaBilledPeriod, FacturaConfig, FacturaData
from pvpcbill.official import (
    DEFAULT_ALQUILER_CONT_ANUAL,
    DEFAULT_BONO_SOCIAL,
    DEFAULT_CUPS,
    DEFAULT_IMPUESTO_ELECTRICO,
    DEFAULT_POTENCIA_CONTRATADA_KW,
    TaxZone,
    TipoPeaje,
)
from pvpcbill.text_bill import bill_text_repr


class FacturaElec:
    """Cálculo de la facturación eléctrica en España para particulares con PVPC."""

    consumo_horario: pd.Series
    pvpc_data: pd.DataFrame
    data: FacturaData

    def __init__(
        self,
        consumo_horario: pd.Series,
        pvpc_data: pd.DataFrame,
        tipo_peaje="GEN",
        potencia_contratada=DEFAULT_POTENCIA_CONTRATADA_KW,
        zona_impuestos="IVA",
        alquiler_anual=DEFAULT_ALQUILER_CONT_ANUAL,
        con_bono_social=DEFAULT_BONO_SOCIAL,
        cups=DEFAULT_CUPS,
        impuesto_electrico=DEFAULT_IMPUESTO_ELECTRICO,
    ):
        # Datos de Consumo y PVPC
        self.pvpc_data = pvpc_data
        self.consumo_horario = consumo_horario

        # Datos de facturación
        initial_config = FacturaConfig(
            tipo_peaje=TipoPeaje(tipo_peaje),
            potencia_contratada=potencia_contratada,
            con_bono_social=con_bono_social,
            zona_impuestos=TaxZone(zona_impuestos),
            alquiler_anual=alquiler_anual,
            cups=cups,
            impuesto_electrico=impuesto_electrico,
        )

        # PROCESADO DE FACTURA
        self._evaluate_bill(initial_config)

    def _evaluate_bill(self, config: FacturaConfig) -> FacturaData:
        """Método para re-generar el cálculo de la factura eléctrica."""
        # Datos de entrada e intervalo
        t0 = self.consumo_horario.index[0]  # .tz_localize(None) - pd.Timedelta("1D")
        tf = self.consumo_horario.index[-1]
        n_days = (tf - t0.replace(hour=0)).days + 1

        # Extrae TCU para tarifa seleccionada de PVPC data
        code = config.tipo_peaje.value
        s_tcu = self.pvpc_data.eval(f"({code} - TEU{code}) / 1000.0")

        # Cálculo de intervalos de facturación:
        periodos_fact = [
            FacturaBilledPeriod.from_hourly_data(
                consumo=self.consumo_horario[self.consumo_horario.index.year == year],
                pvpc_tcu=s_tcu[self.consumo_horario.index.year == year],
                potencia_contratada=config.potencia_contratada,
                tipo_peaje=config.tipo_peaje,
            )
            for year in self.consumo_horario.index.year.astype("category").categories
        ]

        # Init datos de cálculo
        self.data = FacturaData(
            config=config,
            num_dias_factura=n_days,
            start=t0.to_pydatetime(),
            end=tf.to_pydatetime(),
            periodos_fact=periodos_fact,
        )
        return self.data

    ##############################################
    #       Representación                       #
    ##############################################
    def __repr__(self):
        """Representación en texto de la factura eléctrica."""
        return bill_text_repr(self.data)

    def to_dict(self):
        """Representación como `dict` de los componentes de la factura eléctrica."""
        return self.data.to_dict()


pvpcbill/official.py
# -*- coding: utf-8 -*-
"""
Electrical billing for small consumers in Spain using PVPC. Official data.

Facturación de datos de consumo eléctrico para particulares según PVPC.

* Documentos de origen con los valores de los peajes de acceso
  a baja tensión (≤1kV) y baja potencia (≤10kW):
  - [Tarifas 2015 (IDAE)](http://www.idae.es/uploads/documentos/documentos_Tarifas_Reguladas_ene_2015_9195098b.pdf)
  - [Tarifas 2016 (IDAE)](http://www.idae.es/uploads/documentos/documentos_Tarifas_Reguladas_ene_2016_a197c904.pdf)

* Periodo de facturación:
  A partir de 31/03/2014.
  El día de lectura inicial está excluido
  y el día de lectura final está incluido.

* Discriminación horaria:

  - '2.0DHA': Potencia máxima 10 kW, con dos periodos:
    P1 (Punta) Invierno: 12-22h / Verano: 13-23h
    P2 (Valle) Invierno: 22-12h / Verano: 23-13h.

  - '2.0DHS': Tarifa vehículo eléctrico. Potencia máxima 10 kW, con
    P1 (Punta) 13-23h | P2 (Valle) 23-01h / 07-13h | P3 (Supervalle) 01-07h.

  - El horario Valle de 14 horas abarca
    de 22 a 12 h en invierno y de 23 a 13 h en verano.
  - El horario Punta de 10 horas abarca
    de 12 a 22 h en invierno y de 13 a 23 h en verano.
"""  # noqa
from decimal import Decimal, ROUND_HALF_UP
from enum import Enum
from typing import Tuple

import pandas as pd
import pytz
from aiopvpc import ESIOS_TARIFFS

# Defaults y definiciones
DEFAULT_CUPS = "ES00XXXXXXXXXXXXXXSN"
DEFAULT_POTENCIA_CONTRATADA_KW = 3.45
DEFAULT_BONO_SOCIAL = False

ROUND_PREC = 2  # 0,01 €
DEFAULT_IMPUESTO_ELECTRICO = 0.0511269632  # 4,864% por 1,05113
DEFAULT_ALQUILER_CONT_ANUAL = 0.81 * 12  # € / año para monofásico

KEY_TARIFF_GEN = ESIOS_TARIFFS[0]
KEY_TARIFF_NOC = ESIOS_TARIFFS[1]
KEY_TARIFF_VHC = ESIOS_TARIFFS[2]


class TipoPeaje(Enum):
    """Tariff enumerator"""

    GEN = KEY_TARIFF_GEN
    NOC = KEY_TARIFF_NOC
    VHC = KEY_TARIFF_VHC

    @property
    def num_periods(self) -> int:
        if self == self.GEN:
            return 1
        elif self == self.NOC:
            return 2
        return 3

    @property
    def pretty_name(self) -> str:
        if self == self.GEN:
            return "General"
        elif self == self.NOC:
            return "Nocturna"
        return "Vehículo eléctrico"

    @property
    def code(self) -> str:
        if self == self.GEN:
            return "2.0A"
        elif self == self.NOC:
            return "2.0DHA"
        return "2.0DHS"


class TaxZone(Enum):
    """Tax zones enumerator"""

    PENINSULA_BALEARES = "IVA"
    CANARIAS = "IGIC"
    CEUTA_MELILLA = "IPSI"

    @property
    def tax_rate(self) -> float:
        if self == self.PENINSULA_BALEARES:
            return 0.21
        elif self == self.CANARIAS:
            return 0.03
        return 0.01

    @property
    def measurement_tax_rate(self) -> float:
        if self == self.PENINSULA_BALEARES:
            return 0.21
        elif self == self.CANARIAS:
            return 0.07
        return 0.04

    @property
    def pretty_name(self) -> str:
        if self == self.PENINSULA_BALEARES:
            return f"Península y Baleares ({self.value})"
        elif self == self.CANARIAS:
            return f"Canarias ({self.value})"
        return f"Ceuta y Melilla ({self.value})"


# TODO cambiar distinción tarifaria de 'año' a periodo reglamentario
MARGEN_COMERC_EUR_KW_YEAR_MCF = {
    2014: 4.0,  # € /(kW·año)
    2015: 4.0,  # € /(kW·año)
    2016: 4.0,  # € /(kW·año)
    2017: 4.0,  # € /(kW·año)
    # ...
    # TODO add recent data
    2019: 3.113,  # € /(kW·año)
    2020: 3.113,  # € /(kW·año)
}
TERM_POT_PEAJE_ACC_EUR_KW_YEAR_TPA = {
    2014: 35.648148,
    2015: 38.043426,  # 3,503618833 * 12 - 4
    2016: 38.043426,  # (3,1702855 + 0,33333) * 12
    2017: 37.156426,
    # TODO add recent data
    # 2020: 37.258224, + 6c
    # 2020: 37.054905,  # - 4c
    2019: 38.043426,  # ?
    2020: 38.043426,  # +54c 15.85 vs 14.34 + 1.17 = 15.51
    # 2020: 37.156426, # ~ 1 cent (15.52),
    # TODO review period split + round kWh -> 0.103944 + 0.008505 €/kW/dia
}
TERM_ENER_PEAJE_ACC_EUR_KWH_TEA = {
    2014: {
        KEY_TARIFF_GEN: [0.044027],
        KEY_TARIFF_NOC: [0.062012, 0.002215],
        KEY_TARIFF_VHC: [0.074568, 0.017809, 0.006596],
    },
    2015: {
        KEY_TARIFF_GEN: [0.044027],
        KEY_TARIFF_NOC: [0.062012, 0.002215],
        KEY_TARIFF_VHC: [0.074568, 0.017809, 0.006596],
    },
    2016: {
        KEY_TARIFF_GEN: [0.044027],
        KEY_TARIFF_NOC: [0.062012, 0.002215],
        KEY_TARIFF_VHC: [0.062012, 0.002879, 0.000886],
    },
    2017: {
        KEY_TARIFF_GEN: [0.044027],
        KEY_TARIFF_NOC: [0.062012, 0.002215],
        KEY_TARIFF_VHC: [0.062012, 0.002879, 0.000886],
    },
    # TODO add/review recent data
    2019: {
        KEY_TARIFF_GEN: [0.044027],
        KEY_TARIFF_NOC: [0.062012, 0.002215],
        KEY_TARIFF_VHC: [0.062012, 0.002879, 0.000886],
    },
    2020: {
        KEY_TARIFF_GEN: [0.044027],
        KEY_TARIFF_NOC: [0.062012, 0.002215],
        KEY_TARIFF_VHC: [0.062012, 0.002879, 0.000886],
    },
}


##############################################
#       Cálculo                              #
##############################################
def round_money(value: float):
    """Rounding method used in official billing (as it appears)."""
    return float(Decimal(str(value)).quantize(Decimal("1.11"), rounding=ROUND_HALF_UP))


# TODO review sum_money
def round_sum_money(values):
    return sum(round(value, ROUND_PREC) for value in values)


def split_in_tariff_periods(
    series: pd.Series, tipo_peaje: TipoPeaje
) -> Tuple[pd.Series, ...]:
    if tipo_peaje == TipoPeaje.GEN:
        assert tipo_peaje.num_periods == 1
        return (series,)

    # split using constant hours in UTC :)
    idx_utc: pd.DatetimeIndex = series.index.tz_convert(pytz.UTC)
    params_cierre = {"include_start": True, "include_end": False}

    if tipo_peaje == TipoPeaje.NOC:
        idx_p1 = idx_utc.indexer_between_time("11:00", "21:00", **params_cierre)
        idx_p2 = idx_utc.indexer_between_time("21:00", "11:00", **params_cierre)
        assert tipo_peaje.num_periods == 2
        return series[idx_p1], series[idx_p2]

    idx_p1 = idx_utc.indexer_between_time("11:00", "21:00", **params_cierre)
    idx_p2 = idx_utc.indexer_between_time("05:00", "11:00", **params_cierre)
    idx_p3 = idx_utc.indexer_between_time("21:00", "05:00", **params_cierre)
    assert tipo_peaje.num_periods == 3
    return series[idx_p1], series[idx_p2], series[idx_p3]
